# Manual for SGV-caller, SARS-CoV-2 genome variation caller
Jiaqi Wu and So Nakagawa

## 1. Overview
SGV-caller is software that can generate the local database on the genomic variations of SARS-CoV-2 genomes at amino acid, codon, and nucleotide levels. The SGV-caller was designed to handle SARS-CoV-2 genomes and metadata files in the [GISAID database](https://www.gisaid.org/). In particular, it is easy to follow the update of the GISAID database. The output files generated by the SGV-caller are tab-limited text formats that are easy to read, manage and use. Data quality check of each genome is conducted, and the related summary statistics are reported. Data are saved in different format, which are easy access by R software, bash commands as well as other programming languages. Based on the outputs, it will be easy to manage variant information of SARS-CoV-2 genomes and extract the data that meet the user’s requirement.

## 2. Dependency
  The SGV-caller software in this pipeline is using the Perl language. Please make sure you have Perl installed on your computer. In addition, multiple alignment software, MAFFT (1), is required for the software. The installation of the latest version of MAFFT is recommended.
  For Ubuntu user, the easiest way to install MAFFT is as follows:
```
sudo apt-get -y install mafft
```
If you use Mac, you can use [brew](https://brew.sh/) to install it:
```
brew install mafft
```
If you want to install using the source code of the MAFFT suite, you can install it by following the guidance of [MAFFT](https://mafft.cbrc.jp/). If this is the case, please set the path to the binary file of MAFFT programs by:
```
export PATH=$PATH:/path/to/the/directory/where/MAFFT/installed
```
## 3. Instalment 
  Please download the SGV-caller from https://github.com/wujiaqi06/SGV-caller. You can also get it by running: 
```
git clone https://github.com/wujiaqi06/SGV-caller.git
```
  The SGV-caller is Perl software, which does not need compilation. First, you have to set the path to the directory of the SGV-caller as follows:
```
chmod 777 /path/to/the/directory/where/sgv-caller/installed/sgv-caller.pl
export PATH=$PATH:/path/to/the/directory/where/sgv-caller/installed
```
  To use the SGV-caller, you have to copy the SGV-caller’s configuration file “sgv-caller.conf” into the directory containing the data to be analyzed. Then, set the suitables variables in “sgv-caller.conf”, and you can run SGV-caller by the following command:
```
sgv-caller.pl
```
## 4. What can it do
  The main purpose of SGV-caller is to generate the local database of genomic variations of SARS-CoV-2 using data downloaded from [GISAID database](https://www.gisaid.org/). Variations at the nucleotide, amino acid and codon level will be reported. It can be used to user-defined virus genome as well.

## 5. How to run it
  Firstly, please set path to the folder of SGV-caller by: 
```
export PATH=$PATH:Path_to_sgv_caller
```
  Then, copy configuration file “sgv-caller.conf” to the directory where you hope to generate the database, then modify the values of “sgv-caller.conf” file to tell software which calculation you hope to conduct and their input files, then run this in command line:
```
cd where/sgv-caller.conf/is/stored
sgv-caller.pl
```
  By default, SGV-caller will search for the configuration file “sgv-caller.conf” in the database directory. If your configuration file has a different name, please run this command instead, then you can use your own configuration file:
```
sgv-caller.pl -i your_own_configuration_file
```
## 6. Configuration file
  Every time when you run SGV-caller, software firstly reads the configuration file “sgv-caller.conf” for the necessary information it needs. calculation 1-8 in “sgv-caller.conf” corresponding to the pipelines 1-8, which have different functions as follows:
```
Pipeline 1. Making a SGV database start from FASTA sequence and metadata downloaded from GISAID. 
Pipeline 2. Updating existing database with newly downloaded data from GISAID. 
Pipeline 3. Making an SGV database with a given FASTA file, using sequence names as sequence ID.
Pipeline 4. Making an SGV database with the “raw_variants.for_each.all.txt” file.
Pipeline 5. Extracting a subset of the SGV database based on the selected ID.
Pipeline 6. Extracting a FASTA sequences based on a given a GISAID ID list file.
Pipeline 7. Extracting genes from a list of GISAID genomes based on a given GISAID ID list file.
Pipeline 8. Extracting the amino acid replacement information from protein sequences.
```
### 6.1 output_file_name
output_file_name is the run name, which is necessary for pipelines 1-8. It should be a string with only characters, numbers and underbar. Any types of space are not allowed.

### 6.2 reference genome related options
Pipelines 1-5 needs reference genome, as well as its annotation information to annotate the nucleotide variations to codon and amino acid level. By default, SGV-caller takes Wuhan-hu-1 (RefSeq: NC_045512.2) as reference genome. use_default_reference is asking whether you will use the default reference genome or not. It takes the value of “yes” or “no”. If use_default_reference = yes, the default reference genome Wuhan-hu-1 will be used. If use_default_reference = no, the following three files are necessary to run SGV-caller: 
```
custom_reference_genome
custom_codon_fasta_annotation_file
custom_rna_annotation
```
custom_reference_genome is the directory and file name of the custom reference genome file. This file should contain a single sequence, and its sequence name had better not overlap with the sequence name in the fasta data file.
custom_codon_fasta_annotation_file is the directory and file name of the custom codon annotation file. Its format is:
```
S	21563..25384
ORF3a	25393..26220
E	26245..26472
M	26523..27191
nsp10	13025..13441
nsp11	13442..13480
nsp12	13442..13468,13468..16236
…
```
	Comma can be used to separate different genetic regions.
custom_rna_annotation is the directory and file name of the custom RNA annotation file. Its format is:
```
5UTR	1..265
3UTR	29675..29903
Non_Coding_ORF1ab_S	21556..21562
Non_Coding_S_ORF3a	25385..25392
…
```


#######
Text that is not a quote
> Text that is a quote
Some basic Git commands are:
```
git status
git add
git commit
```

This site was built using [GitHub Pages](https://pages.github.com/).

